src/mapreduce.jl

```diff
@@ -90,6 +90,36 @@ end
     end
 end
 
+@generated function reduce{D}(op, a::StaticArray, ::Type{Val{D}})
```

Comment:
I didn't know about `Base.reducedim` either.  Thanks for pointing this out!

To mimic `Base.reducedim` more closely, we should be able to pass a vector of dimensions (rather than a single dimension) in which reduction occurs.  This means that we would need to pass a vector of `Val{D}`'s, which requires the function to be parametrized as `reducedim{D1,D2,...,Dn}` for an arbitrary `n`.

Is specifying an arbitrary number of parameters possible in Julia?  Or we just opt with a single dimension?